include/mbedtls/pkcs11_client.h

```diff
@@ -0,0 +1,150 @@
+/**
+ * \file pkcs11_client.h
+ *
+ * \brief Generic wrapper for Cryptoki (PKCS#11) support
+ *
+ *  Copyright (C) 2017, ARM Limited, All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  This file is part of mbed TLS (https://tls.mbed.org)
+ */
+#ifndef MBEDTLS_PKCS11_CLIENT_H
+#define MBEDTLS_PKCS11_CLIENT_H
+
+#if !defined(MBEDTLS_CONFIG_FILE)
+#include "config.h"
+#else
+#include MBEDTLS_CONFIG_FILE
+#endif
+
+#if defined(MBEDTLS_PKCS11_CLIENT_C)
+
+#include <pkcs11.h>
+
+#if ( defined(__ARMCC_VERSION) || defined(_MSC_VER) ) && \
+    !defined(inline) && !defined(__cplusplus)
+#define inline __inline
+#endif
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define MBEDTLS_PKCS11_FLAG_TOKEN    ( (uint32_t) 0x80000000 )
+
+#if defined(MBEDTLS_PK_C)
+
+#define MBEDTLS_PK_FLAG_SENSITIVE    ( (uint32_t) 0x00000001 )
+#define MBEDTLS_PK_FLAG_EXTRACTABLE  ( (uint32_t) 0x00000002 )
+#define MBEDTLS_PK_FLAG_SIGN         ( (uint32_t) 0x00000010 )
+#define MBEDTLS_PK_FLAG_VERIFY       ( (uint32_t) 0x00000020 )
+#define MBEDTLS_PK_FLAG_DECRYPT      ( (uint32_t) 0x00000040 )
+#define MBEDTLS_PK_FLAG_ENCRYPT      ( (uint32_t) 0x00000080 )
+
+#include "pk.h"
+
+/**
+ * \brief               Set up a PK context for a key pair in a PKCS#11 token
+ *
+ * \param ctx           PK context to fill, which must have been initialized
+ *                      with mbedtls_pk_init().
+ * \param hSession      Cryptoki session.
+ * \param hPublicKey    Cryptoki handle of the public key.
+ * \param hPrivateKey   Cryptoki handle of the private key, or
+ *                      CK_INVALID_HANDLE for a public key rather than a key
+ *                      pair.
+ *
+ * \return              0 on success,
+ *                      or MBEDTLS_ERR_PK_XXX error code.
+ *
+ * \note                The session and the key(s) must remain valid until the
+ *                      PK context is closed with mbedtls_pk_free(). As an
+ *                      exception, it's ok to call mbedtls_pk_free() itself
```

Comment:
I am not sure I understand this sentence, could you please clarify?

Reply:
Under the hood, operations using the pk object use the session handle and the key handle. Therefore these handles must not be closed while the pk object exist. For me this is clear from the wording in the note, but if it isn't clear to you then we should clarify. What do you think of adding my first two sentences here?

Reply:
Thank you for the clarification, but this was clear from the first sentence (lines 72-73). For me the second sentence (lines 74-75) seems to contradict the first one. How can it be ok to call `mbedtls_pk_free()` when the Cryptoki handles have become invalid, if they must remain valid until `mbedtls_pk_free()` is called?

Reply:
@yanesca The handles are needed by most functions that use the `pk` object, but not by `mbedtls_pk_free`. `mbedtls_pk_free` does not contact the PKCS#11 token. It merely erases the reference to the token from the object in mbedtls's memory.

Another way to phrase it is to use the contraposite: if any of the handles become invalid, then you may no longer do anything with the pk object except call `mbedtls_pk_free` on it.

The advantage of stating in the documentation that `mbedtls_pk_free` doesn't need the handles is that it gives more freedom to the application. For example, if it notices that the token has disconnected, it can close the session before freeing the pk objects.

Arguably a better interface would manage the connection to the token under the hood and would not expose PKCS#11 handles at all. But that means transferring the responsibility of opening the key inside Mbed TLS, and that often requires implementation-specific function calls, so it isn't always possible.

Reply:
Thanks for the clarification. Putting it this way is much clearer to me: ` if any of the handles become invalid, then you may no longer do anything with the pk object except call mbedtls_pk_free on it.`

Reply:
Done.